### ISSUE TITLE
[BUG] Storage doesn't reschedule in v1.7.2

### ISSUE THREAD

**Issue by cranky-coder** [longhorn/longhorn#10090]
## Describe the bug

Since upgrading to 1.7.2 when a workload gets moved to another node, the storage does not get moved with it.  I have had longhorn running for quite some time with anywhere between 2 and 6 replicas and things for the most part worked.  But since upgrading to 1.7.2 I can scale a workload to 0.  Go into the UI and force a detach and the volume will automatically reattach.  If I try to start the workload if it's on a different node then where the volume attached it fails.

## Expected behavior

When a workload is rescheduled the storage should reschedule as well.

## Environment

 - Longhorn version: 1.7.2
 - Impacted volume (PV): all of them since upgrade
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): kubectl
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: k3s
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 6
 - Node config
   - OS type and version: arm64/amd64
   - Kernel version: 
   - CPU per node: varies
   - Memory per node: varies
   - Disk type (e.g. SSD/NVMe/HDD): ssd
   - Network bandwidth between the nodes (Gbps): gig
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): baremetal
 - Number of Longhorn volumes in the cluster: 24

## Workaround and Mitigation

Have tried modifying the volumeattachments and removing the finalizers, have tried clearing the spec.nodeID.

**Comment by innobead** [comment]
Could be related to https://github.com/longhorn/longhorn/issues/9934

Can u provide the support bundle? 

cc @derekbit @PhanLe1010 @c3y1huang

**Comment by innobead** [comment]
Have u tried https://github.com/longhorn/longhorn/issues/9781#issuecomment-2475753876?

**Comment by cranky-coder** [comment]
> Could be related to #9934
> 
> Can u provide the support bundle?
> 
> cc @derekbit @PhanLe1010 @c3y1huang

Trying.  Longhorn UI stuck on generating bundle.

**Comment by cranky-coder** [comment]
> Have u tried [#9781 (comment)](https://github.com/longhorn/longhorn/issues/9781#issuecomment-2475753876)?

![image](https://github.com/user-attachments/assets/441742eb-c37d-4bfb-b4d5-20ef95362bd1)

I have node soft affinity enabled, but so are the others.  Should I isolate just node?

**Comment by innobead** [comment]
No need, but have you successfully generated the support bundle?

https://longhorn.io/docs/1.7.2/troubleshoot/support-bundle/
https://longhorn.io/kb/troubleshooting-create-support-bundle-with-curl/

**Comment by cranky-coder** [comment]
Support bundle is 155meg.  Is there somewhere else i can upload?

**Comment by derekbit** [comment]
> Support bundle is 155meg. Is there somewhere else i can upload?

You can upload to your cloud storage and share us with the link

**Comment by cranky-coder** [comment]
support bundle 

https://drive.google.com/file/d/1VfbWDjQSGq1JyPxd34NfNJ2TSxpYU-oT/view?usp=sharing

**Comment by derekbit** [comment]
> support bundle
> 
> [drive.google.com/file/d/1VfbWDjQSGq1JyPxd34NfNJ2TSxpYU-oT/view?usp=sharing](https://drive.google.com/file/d/1VfbWDjQSGq1JyPxd34NfNJ2TSxpYU-oT/view?usp=sharing)

Please provide the name of the problematic volume.

**Comment by cranky-coder** [comment]
its most of them.  but for now use this one specifically.

pvc-72d8c542-4d60-4e99-bac3-693f49a6ee67

**Comment by cranky-coder** [comment]
Was anything jumping out in the support bundle.  I left some of the workloads alone and have given it time to see if maybe something in the cluster just needed to settle, but the storage is still just staying where it was.

**Comment by derekbit** [comment]
@cranky-coder 
I'm on vacation and sorry for the late reply. Longhorn's volumeattachment is used by backup jobs, so the CSI is unable to attach the volume. However, the backup jobs cannot find the snapshots for backup and always retry. It seems related to  https://github.com/longhorn/longhorn-manager/pull/2835. We need to further investigate it.

Current workaround is to delete the `Backup` resources whose `status.state` are `pending` and `status.messages` is `Failed to get the Snapshot...`.

cc @ChanYiLin

**Comment by derekbit** [comment]
@ChanYiLin will be off for one week. @PhanLe1010 Can you help check the issue?  Thank you.

**Comment by derekbit** [comment]
Three points
1. volumeattachment preemption 
2. why is snapshot not created
3. max retires of the failed backup

**Comment by cranky-coder** [comment]
Deleted backup's with state.status=pending.  That helped alot!!! What's odd though is I am able to see the snapshots and backups in the UI.  Is there something different that isn't seeing them?  I just have it using s3 object store to a minio setup.  

Or is this potentially an issue with pending backups on the old version prior to upgrade?  Is there something permanent I should do to prevent the backup's from locking up my volumes like this?

**Comment by derekbit** [comment]
cc @PhanLe1010

**Comment by PhanLe1010** [comment]
Hi @cranky-coder What was the version before upgrading to version v1.7.2?

**Comment by PhanLe1010** [comment]
Tracking the volume `pvc-72d8c542-4d60-4e99-bac3-693f49a6ee67`. Looks like the `backup-*` tickets are not the blocker. The direct reason it cannot attach to new node `kube-8` is because there are other `csi-*` tickets already request the volume to attach to other nodes and this volume is RWO volume:
```yaml
      csi-59b518f1ceddffe2a7fab4e81bca62e5f0f4dbe116d345895ee670c17a2e1e71:
        conditions:
        - lastProbeTime: "null"
          lastTransitionTime: "2024-12-27T00:14:43Z"
          message: 'the volume is currently attached to different node kube-8 '
          reason: "null"
          status: "False"
          type: Satisfied
        generation: 0
        id: csi-59b518f1ceddffe2a7fab4e81bca62e5f0f4dbe116d345895ee670c17a2e1e71
        satisfied: false
      csi-43391ce9dd758bc3242b0a5fba33f97d8e55023ea6a7c38eb7f838a9435c84cb:
        conditions:
        - lastProbeTime: "null"
          lastTransitionTime: "2024-12-27T06:14:25Z"
          message: "null"
          reason: "null"
          status: "True"
          type: Satisfied
        generation: 0
        id: csi-43391ce9dd758bc3242b0a5fba33f97d8e55023ea6a7c38eb7f838a9435c84cb
        satisfied: true
      csi-c6669469a7646c5d9e547e56a6bb002dd16129fa97201bb41b62415501fe909e:
        conditions:
        - lastProbeTime: "null"
          lastTransitionTime: "2024-12-27T06:15:11Z"
          message: 'the volume is currently attached to different node kube-8 '
          reason: "null"
          status: "False"
          type: Satisfied
        generation: 0
        id: csi-c6669469a7646c5d9e547e56a6bb002dd16129fa97201bb41b62415501fe909e
        satisfied: false
```

The question is why are there extra CSI AD ticket like this? By design, the `csi-*` tickets should corresponding to the CSI VA CR. However, there is currently only CSI VA CR at the moment on `node-8`:
```yaml
    name: csi-c6669469a7646c5d9e547e56a6bb002dd16129fa97201bb41b62415501fe909e
    resourceVersion: "1062285481"
    uid: 68b09531-edba-4a91-860c-7a98fcd0aac1
  spec:
    attacher: driver.longhorn.io
    nodeName: kube-7
    source:
      persistentVolumeName: pvc-72d8c542-4d60-4e99-bac3-693f49a6ee67
  status:
    attachError:
      message: 'rpc error: code = Internal desc = volume pvc-72d8c542-4d60-4e99-bac3-693f49a6ee67
        failed to attach to node kube-7 with attachmentID csi-c6669469a7646c5d9e547e56a6bb002dd16129fa97201bb41b62415501fe909e:
        the volume is currently attached to different node kube-8 '
```

So, there is an out of sync between CSI VA CR and Longhorn CSI tickets here

**Comment by PhanLe1010** [comment]
Will continue on this tomorrow

**Comment by PhanLe1010** [comment]
One quick question @cranky-coder 

Continue on the issue 1 in the comment https://github.com/longhorn/longhorn/issues/10090#issuecomment-2574273380, the situation looks as if someone/something manually delete the CSI VA CR (`volumeattachments.storage.k8s.io`) and remove its finalizer directly without waiting for longhorn-csi plugin to clean up the `csi-*` tickets. Do you or your script happen to do this by any chance?

**Comment by PhanLe1010** [comment]
Thanks @cranky-coder for the additional information.

## Issue 1
I think the story for issue 1 might be like the one below. This should be expected due to the manual intervention and I can reproduce similar behavior.
1. There are some backups of the volume which cannot find the corresponding snapshot (this is the issue 2 which will be discussed below)
1. These backup CRs hold the `backup-*` tickets inside the Longhorn VA CR 
1. When the user scales down the workload, the volume does not detach because of these `backup-*` tickets
1. So the user thinks that the volume is stuck on the current note. However, note that the volume is NOT actually stuck on this node because if the user scales up the workload pod, the `csi-*` ticket will have higher priority than the `backup-*` tickets and therefore will make the volume attach to the node of the workload pod
1. The users tried to manually fix the problem for other volumes by manually removing finalizer of CSI VA CRs and scale down the workload pods of these volumes. 
1. This leads to CSI VA CR and Longhorn VA CA is out of sync. A.K.A, there are `csi-*` ticket inside Longhorn VA CR but there is no corresponding CSI VA CR
1. When user scale up the workload pod, kube controller mamanager create new CSI VA CR. Longhorn create new `csi-*` ticket. However, Longhorn cannot attach the volume to the node of the new `csi-*` ticket because there are old out-of-sync `csi-*` ticket on other nodes.

In summary, I think the 2 main points of this issue is:
1. The confusion that the volume is stuck because of the `backup-*` tickets. The correct understanding should be even though the volume is attached due to `backup-*` tickets, it can still be interrupted and attach to new node once the workload pod come up and create `csi-*` ticket.
1. Manually remove finalizer and delete CSI VA CR might make them out of sync with Longhorn VA CR. This will cause the volume to actually stuck and cannot move to new nodes. The workaround once the user hit this state is scale down all workload pod, force detach the volume using UI, scale up the workload pod

## Issue 2:
... Updating

**Comment by cranky-coder** [comment]
I have this happening right now with a workload.  I have not touched it in case there is any troubleshooting we can do.  

Workload moved to kube-2.  Error is saying volume is already attached to kube-7.  Checking volumeattachments only shows the one trying t mount to kube-2.  Longhorn UI says it's attached to kube-7 as well.  Previous pod has been fully terminated.  Is there something I can look for to see if the longhorn finalizer is waiting for something?  I will leave this one as is without touching it and see if it recovers on it's own or if you would like another support bundle.

**Comment by PhanLe1010** [comment]
Hi @cranky-coder 

What is the name of the volume? 
Can you send me a support bundle taken while the issue is happening?

**Comment by cranky-coder** [comment]
Support bundle: https://drive.google.com/file/d/1L4YQ-3_8nZE8ApXNQOWFLj8Q1EeNhmTL/view?usp=drive_link
pod name/status: overseer-6664475ff4-dtssj / "container creating"

error: Warning  FailedAttachVolume  2m33s (x1303 over 44h)  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-002ff1ae-d62e-4671-8a7c-cedbbdf4ee0c" : rpc error: code = Internal desc = volume pvc-002ff1ae-d62e-4671-8a7c-cedbbdf4ee0c failed to attach to node kube-2 with attachmentID csi-74a1e022950caf7f015ddc2f2b75007d611ef8948d4e9cf96f4f2b1a1d688f80: the volume is currently attached to different node kube-7

volume name: pvc-002ff1ae-d62e-4671-8a7c-cedbbdf4ee0c

Screenshot of volume attachment for that PV
![image](https://github.com/user-attachments/assets/0f3899ed-1e34-411b-9a6e-fa7ab9170b6a)

Issue is happening currently.  Haven't touched the pod so this is current state.

**Comment by cranky-coder** [comment]
Granted

**Comment by PhanLe1010** [comment]
Hi @cranky-coder There is an orphan `csi-*` ticket which might be the result from manual actions in the past. 
More specifically, the ticket `csi-4349cfe336ae35a31ead12e6b3ee9b61f8182aff782262ca6b3e017d03ccb14c` exist in `volumeattachment.longhorn.io` but there is no corresponding `VolumeAttachment.storage.k8s.io` object
```yaml
apiVersion: v1
items:
- apiVersion: longhorn.io/v1beta2
  kind: VolumeAttachment
  metadata:
    creationTimestamp: "2024-10-28T14:47:56Z"
    finalizers:
    - longhorn.io
    labels:
      longhornvolume: pvc-002ff1ae-d62e-4671-8a7c-cedbbdf4ee0c
    name: pvc-002ff1ae-d62e-4671-8a7c-cedbbdf4ee0c
    namespace: longhorn-system
  spec:
    attachmentTickets:
      csi-74a1e022950caf7f015ddc2f2b75007d611ef8948d4e9cf96f4f2b1a1d688f80:
        generation: 0
        id: csi-74a1e022950caf7f015ddc2f2b75007d611ef8948d4e9cf96f4f2b1a1d688f80
        nodeID: kube-2
        parameters:
          disableFrontend: "false"
          lastAttachedBy: "null"
        type: csi-attacher
      csi-4349cfe336ae35a31ead12e6b3ee9b61f8182aff782262ca6b3e017d03ccb14c:
        generation: 0
        id: csi-4349cfe336ae35a31ead12e6b3ee9b61f8182aff782262ca6b3e017d03ccb14c
        nodeID: kube-7
        parameters:
          disableFrontend: "false"
          lastAttachedBy: "null"
        type: csi-attacher
    volume: pvc-002ff1ae-d62e-4671-8a7c-cedbbdf4ee0c
```

You can clean them up by:
1. Scale down the workload
1. Force detach volume volume OR edit the `volumeattachment.longhorn.io` CR to remove the ticket `csi-4349cfe336ae35a31ead12e6b3ee9b61f8182aff782262ca6b3e017d03ccb14c`
1. Scale up the workload

**Comment by PhanLe1010** [comment]
Reproducing and test plan:
1. Create a Longhorn volume
1. Deploying a backup with non-existing snapshot
    ```yaml
    apiVersion: longhorn.io/v1beta2
    kind: Backup
    metadata:
      name: test-backup
      namespace: longhorn-system
      labels:
        backup-volume: pvc-34e83a41-bb6a-4406-b9d9-91fe3a8fcfbc # Replace by the name of your volume
    spec:
      backupMode: incremental
      snapshotName: non-exist-snapshot-name
    ```
1. Before the fix, the backup will stuck in pending state forever and keep the volume in attached state
1. After the fix, the backup is marked as error and the volume is detached after around 2 minutes

**Comment by yangchiu** [comment]
>     4. After the fix, the backup is marked as error and the volume is detached after around 2 minutes

Hi @PhanLe1010, does that mean the volume will be detached automatically? I tested on v1.8.x-head (longhorn-manager 2c33bee) with the following steps:

1. Create a workload for a volume:
```
apiVersion: v1
kind: Service
metadata:
  name: mysql-2
  labels:
    app: mysql-2
spec:
  ports:
    - port: 3306
  selector:
    app: mysql-2
  clusterIP: None
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mysql-pvc-2
  namespace: default
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: longhorn
  resources:
    requests:
      storage: 10Gi
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mysql-2
  labels:
    app: mysql-2
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: mysql-2 # has to match .spec.template.metadata.labels
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: mysql-2
    spec:
      restartPolicy: Always
      containers:
      - image: mysql:5.6
        name: mysql
        livenessProbe:
          exec:
            command:
              - ls
              - /var/lib/mysql/lost+found
          initialDelaySeconds: 5
          periodSeconds: 5
        ports:
        - containerPort: 3306
          name: mysql
        volumeMounts:
        - name: mysql-volume
          mountPath: /var/lib/mysql
        env:
        - name: MYSQL_ROOT_PASSWORD
          value: "rancher"
      volumes:
      - name: mysql-volume
        persistentVolumeClaim:
          claimName: mysql-pvc-2
``` 
2. Create a backup for it with a non-existing snapshot:
```
# create backup from yaml
apiVersion: longhorn.io/v1beta2
kind: Backup
metadata:
  name: test-backup
  namespace: longhorn-system
  labels:
    backup-volume: pvc-39f96b76-06ac-4dad-8b13-a2c5e44c5608 # Replace by the name of your volume
spec:
  backupMode: incremental
  snapshotName: non-exist-snapshot-name
```

I can observe the backup being marked as `Error` after a while:
```
$ kubectl get backups -n longhorn-system -w
NAME          SNAPSHOTNAME   SNAPSHOTSIZE   SNAPSHOTCREATEDAT   BACKUPTARGET   STATE     LASTSYNCEDAT
test-backup                                                                    Pending   <no value>
test-backup                                                                    Error     2025-01-13T01:40:18Z
```
But the volume keeps in `Attached` state indefinitely:
```
$ kubectl get volumes -n longhorn-system -w
NAME                                       DATA ENGINE   STATE      ROBUSTNESS   SCHEDULED   SIZE          NODE            AGE
pvc-39f96b76-06ac-4dad-8b13-a2c5e44c5608   v1            attached   healthy                  10737418240   ip-10-0-1-146   21m
```

Could you help confirm the behavior? Thank you!

**Comment by yangchiu** [comment]
Verification passed on master-head (longhorn-manager 3c46af5) and v1.8.x-head (longhorn-manager 2c33bee) following [the test steps](https://github.com/longhorn/longhorn/issues/10090#issuecomment-2584797283).
